### PR TITLE
XRDDEV-126

### PR DESCRIPTION
**Pull request by jhyoty**

### CODE REVIEW

File: src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java [hunk 1]
```diff
@@ -165,17 +154,85 @@ protected AuthKeyInfo getAuthKeyInfo(SecurityServerId serverId) throws Exception
...
+    static CachingKeyConfImpl newInstance() throws Exception {
+        final FileContentChangeChecker changeChecker = new FileContentChangeChecker(SystemProperties.getKeyConfFile());
+        final CachingKeyConfImpl instance = new CachingKeyConfImpl();
+        // the change watcher can not be created in the constructor, because that would publish the
+        // instance reference to another thread before the constructor finishes.
+        instance.keyConfChangeWatcher = createChangeWatcher(new WeakReference<>(instance), changeChecker);
+        return instance;
+    }
+
+    /* Implementation note:
```
Comment (jansu76) [inline src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java:211]: :+1: nice to have this kind of implementation notes

**Review by jansu76 — CHANGES_REQUESTED** [review]
The getAuthKey optimization seems to be a close call, on whether it is justified. As I understand, the performance benefits are real, but quite small (3% - maybe 6% best case?). As a price for that, they add some amount of complexity to the overall solution. The code changes related to getAuthKey improvements are themselves relatively small and straightforward, but they do add some complexity that people working with key caching and ocsp freshness etc need to understand.

Description at https://confluence.niis.org/pages/viewpage.action?pageId=84837614 sounds correct, but I don't understand all they mechanisms that relate to key validation well enough to validate that the change does not cause any side effects somewhere (sort of related to the previous chapter - the whole is already so complex that I feel timid about adding any new complexity).

File: src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java [hunk 2]
```diff
@@ -165,17 +154,85 @@ protected AuthKeyInfo getAuthKeyInfo(SecurityServerId serverId) throws Exception
...
+                notAfter);
+    }
+
+    // Upper bound for validity is the minimum of certificates "notAfter" and OCSP responses validity time
+    static Date calculateNotAfter(List<OCSPResp> ocspResponses, Date notAfter) throws OCSPException {
+        final long freshnessMillis = 1000L * GlobalConf.getOcspFreshnessSeconds(false);
+        final boolean verifyNextUpdate = GlobalConfExtensions.getInstance().shouldVerifyOcspNextUpdate();
+
+        for (OCSPResp resp : ocspResponses) {
+            final SingleResp singleResp = ((BasicOCSPResp) resp.getResponseObject()).getResponses()[0];
```
[the remark this answers is not in the document]
Comment (jhyoty) [reply src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java:186]: One could request OCSP responses for multiple certificates using just one request (--> multiple responses), but X-Road does not do that.

File: src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java [hunk 3]
```diff
@@ -165,17 +154,85 @@ protected AuthKeyInfo getAuthKeyInfo(SecurityServerId serverId) throws Exception
...
+            final Date tmp = new Date(singleResp.getThisUpdate().getTime() + freshnessMillis);
```
[the remark this answers is not in the document]
Comment (jhyoty) [reply src/proxy/src/main/java/ee/ria/xroad/proxy/conf/CachingKeyConfImpl.java:188]: Perhaps, but given the very narrow context it is used in, I think this is readable enough.

**Review by jansu76 — APPROVED** [review]

**Pull request merged**